greenwave/policies.py

```diff
@@ -471,35 +495,26 @@ def to_json(self):
         }
 
 
-def _summarize_answers_without_errored(answers):
-    failure_count = sum(
-        1 for answer in answers
-        if isinstance(answer, RuleNotSatisfied)
-    )
-    missing_count = sum(
-        1 for answer in answers
-        if isinstance(answer, (TestResultIncomplete, TestResultMissing))
-    )
-
-    # Missing results are also failures but we will distinguish between those
-    # two in summary message.
-    failure_count -= missing_count
-
-    if failure_count and missing_count:
-        return '{} of {} required tests failed, {} result{} missing'.format(
-            failure_count, len(answers), missing_count, 's' if missing_count > 1 else '')
-
-    if failure_count > 0:
-        return '{} of {} required tests failed'.format(failure_count, len(answers))
-
-    if missing_count > 0:
-        return '{} of {} required test results missing'.format(missing_count, len(answers))
-
-    if all(answer.is_satisfied for answer in answers):
-        return 'All required tests passed'
-
-    logging.error('Unexpected unsatisfied result')
-    return 'inexplicable result'
+class _Summary(object):
+    """
+    Internal class using for generating the result summary.
+    """
+    def __init__(self):
+        self.test_msgs = defaultdict(int)
+        self.non_test_msgs = defaultdict(int)
+
+    def to_text(self, test_count):
+        msgstr = ""
+        if self.non_test_msgs:
+            msgstr = ", ".join("{} {}".format(v, k) for (k, v) in self.non_test_msgs.items())
```

Comment:
`msgstr = ", ".join(k.format(num=v, s="s" * {v > 1}) for (k, v) in self.non_test_msgs.items())`